acurl/src/session.pyx

```diff
@@ -0,0 +1,258 @@
+#cython: language_level=3
+
+import time  # FIXME: use fast c fn
+
+from libc.stdlib cimport malloc
+from libc.string cimport strndup
+from cpython.pycapsule cimport PyCapsule_New, PyCapsule_GetPointer
+from curlinterface cimport *
+from cpython.ref cimport Py_INCREF
+from libc.stdio cimport printf
+import cython
+from json import dumps
+
+class RequestError(Exception):
+    pass
+
+# Callback functions
+
+cdef BufferNode* alloc_buffer_node(size_t size, char *data):
+    cdef BufferNode* node = <BufferNode*>malloc(sizeof(BufferNode))
+    if node == NULL:
+        printf("OOOPS!!!!")  # FIXME: better checks
+    node.len = size
+    node.buffer = strndup(data, size)
+    node.next = NULL
+    return node
+
+cdef size_t header_callback(char *ptr, size_t size, size_t nmemb, void *userdata):
+    cdef _Response response = <_Response>userdata
+    Py_INCREF(response)  # FIXME: why
+    cdef BufferNode* node = alloc_buffer_node(size * nmemb, ptr)
+    if response.header_buffer == NULL:  # FIXME: unlikely
+        response.header_buffer = node
+    if response.header_buffer_tail != NULL:  # FIXME: likely
+        response.header_buffer_tail.next = node
+    response.header_buffer_tail = node
+    return node.len
+
+cdef size_t body_callback(char *ptr, size_t size, size_t nmemb, void *userdata):
+    cdef _Response response = <_Response>userdata
+    Py_INCREF(response)  # FIXME: why
+    cdef BufferNode* node = alloc_buffer_node(size * nmemb, ptr)
+    if response.body_buffer == NULL:  # FIXME: unlikely
+        response.body_buffer = node
+    if response.body_buffer_tail != NULL:  # FIXME: likely
+        response.body_buffer_tail.next = node
+    response.body_buffer_tail = node
+    return node.len
+
+cdef void cleanup_share(object share_capsule):
+    cdef void* share_raw = PyCapsule_GetPointer(share_capsule, <const char*>NULL)
+    curl_share_cleanup(<CURLSH*>share_raw)
+
+@cython.no_gc_clear  # FIXME: make sure this isn't going to leak!
+cdef class Session:
+    cdef CURLSH* shared
+    cdef CurlWrapper wrapper
+    cdef public object response_callback
+
+    def __cinit__(self, wrapper):
+        self.shared = curl_share_init()
+        acurl_share_setopt_int(self.shared, CURLSHOPT_SHARE, CURL_LOCK_DATA_COOKIE)
+        acurl_share_setopt_int(self.shared, CURLSHOPT_SHARE, CURL_LOCK_DATA_DNS)
+        acurl_share_setopt_int(self.shared, CURLSHOPT_SHARE, CURL_LOCK_DATA_SSL_SESSION)
+        self.wrapper = wrapper
+        self.response_callback = None
+
+    def __dealloc__(self):
+        if self.wrapper.loop.is_closed():
+            # FIXME: the event loop being closed only happens during testing,
+            # so it kind of sucks to pay the price of checking for it all the
+            # time...
+            curl_share_cleanup(self.shared)
+        else:
+            self.wrapper.loop.call_soon(cleanup_share, PyCapsule_New(self.shared, NULL, NULL))
+
+    def cookies(self):
```

Comment:
Hmm.  This implies that we can extract the cookies from a session without doing a dummy request (contrary to the previous acurl implementation...)  Does it work?

Reply:
It does :tada: This makes some of the things we were talking about the other day significantly easier.

Reply:
Agreed, I stumbled upon this whilst testing after our initial conversation